torch/_subclasses/fake_tensor.py

```diff
@@ -2095,79 +2102,159 @@ def _check_fake_real_vals(fake: Any, real: Any) -> None:
-                                f"fake tensor {_fake_out} and real tensor {_real_out}, "
-                                f" at output index {i}, for func: {func}"
+                                f"fake tensor {fake} and real tensor {real}, "
+                                f" at output{keystr(path)}, for func: {func}"
                             ) from exc
 
                         for j, (s_fake, s_real) in enumerate(
-                            zip(_fake_out.size(), _real_out.size())
+                            zip(fake.size(), real.size())
                         ):
                             try:
                                 _check_fake_real_vals(s_fake, s_real)
                             except MetadataMismatchError as exc:
+                                if (
+                                    torch._functorch.config.generate_fake_kernels_from_real_mismatches
+                                ):
+                                    dtrace_structured(
+                                        "mismatched_fake_kernel",
+                                        metadata_fn=lambda: {
+                                            "op": str(func),
+                                            "reason": exc.reason,
+                                        },
+                                    )
+                                    return _make_fake(self, func, real)
                                 raise MetadataMismatchError(
                                     f"Real tensor propagation found an output size mismatch between "
-                                    f"fake shape {s_fake} and real shape {s_real}, at output "
-                                    f"index {i}, dimension {j} for func: {func}"
+                                    f"fake shape {s_fake} and real shape {s_real}, "
+                                    f"at output{keystr(path)}.size({j}), for func: {func}"
                                 ) from exc
                     else:
                         try:
-                            _check_fake_real_vals(_fake_out, _real_out)
+                            _check_fake_real_vals(fake, real)
                         except MetadataMismatchError as exc:
                             raise MetadataMismatchError(
                                 f"Real tensor propagation found an output value mismatch between "
-                                f"fake output value {_fake_out} and real output value {_real_out}, "
-                                f" at output index {i}, for func: {func}"
+                                f"fake output value {fake} and real output value {real}, "
+                                f"at output{keystr(path)}, for func: {func}"
                             ) from exc
+                    return fake
+
+                # check fake/real tensor properies, sizes & output values
+                fake_paths_leaves, fake_spec = pytree.tree_flatten_with_path(fake_out)
```

Comment:
Can you please factor this out ? this function is already way too long and difficult too follow.